CRM/ManualDirectDebit/Hook/Custom/NecessaryFieldGenerator.php

```diff
@@ -0,0 +1,242 @@
+<?php
+
+/**
+ * Generates necessary field if it wasn't filed by user
+ */
+class CRM_ManualDirectDebit_Hook_Custom_NecessaryFieldGenerator {
+
+  /**
+   * Primary ID of Direct Debit Mandate
+   *
+   * @var int
+   */
+  private $mandateId;
+
+  /**
+   * Contact entity ID
+   *
+   * @var int
+   */
+  private $entityID;
+
+  /**
+   * Array of necessary fields, which must to be generated
+   *
+   * @var array
+   */
+  private $necessaryFields = [
+    'collection_day' => FALSE,
+    'dd_ref' => FALSE,
+    'authorisation_date' => FALSE,
+    'start_date' => FALSE,
+  ];
+
+  /**
+   * Array of extension settings
+   *
+   * @var array
+   */
+  private $settings;
+
+  public function __construct($entityID) {
+    $this->entityID = $entityID;
+    $this->mandateId = $this->getMandateId();
+    $this->settings = $this->getManualDirectDebitSettings();
+  }
+
+  /**
+   * Launches process of necessary field generation
+   *
+   * @param $params
+   */
+  public function create(&$params) {
+    $this->findFieldWhichNeedToBeGenerate($params);
```

Comment:
I would combine both findFieldWhichNeedToBeGenerate and generateValues in one method **generateFieldsValues()**